### ISSUE TITLE
[Core Feature] Wait on ongoing task when cache is enabled

### ISSUE THREAD

**Issue by michaels-lyft** [flyteorg/flyte#872]
**Motivation: Why do you think this is important?**
This would allow all concurrent cached Flyte tasks with the same input to wait on the first one to complete instead of causing a race condition when they execute at the same time. This would also help save resources by preventing repeated work.

**Goal: What should the final outcome look like, ideally?**
When multiple cached tasks with the same inputs are running concurrently, only one should be executing, the rest should be waiting on that one to finish.

**Describe alternatives you've considered**
Writing our own cache system by hashing inputs, storing them in s3, and checking them before executing the main logic.

**[Optional] Propose: Link/Inline OR Additional context**
Cached tasks already exist today, but it only works for already completed tasks. The proposal is to enable it for ongoing tasks as well.

**Comment by kumare3** [comment]
@michaels-lyft thank you for the issue. I have a few questions/comments, inputs will be awesome 
- are these tasks in the same workflow? I am assuming not and you would want to reuse outputs across workflows that may be run concurrently 
- the goal here is to save money, as if they are not concurrent then the cache reuse will work

Some problems with doing a transaction on the cache
- we will need to provide semantics of serialization on that same key on the cache. And most importantly have the right UX to show a potential in progress execution for the same task with matching inputs
- we will need to handle cases in which the previous execution fails, then the - global lock is voided, and to do this correctly we will need some leasing semantics 

Solution 
- a task can have an additional metadata setting that would enable cache serialization and will also have a timeout to wait for the global lock to expire
@task(cache_serialization_timeout=1h,...)

- the Ui should show a message that a matching task is in progress and serialization is enabled

- if the timeout is less than the total execution time, then it will only wait for the timeout and if cache is still not available, it will proceed

@michaels-lyft / @kanterov / @slai what do you guys think about this ^

**Comment by michaels-lyft** [comment]
Thank you for the quick response @kumare3 !  

> are these tasks in the same workflow?

For my use case, it's the same tasks from the same workflows that are being executed by another workflow and are running concurrently. For example, I have workflow A, which has a cached task A, and other subsequent tasks B, C, etc. Then I have workflow B, which calls workflow A multiple times concurrently through a dynamic_task. The multiple concurrent A tasks would all have the same inputs but the subsequent tasks B, C, etc. would all have different inputs.

> the goal here is to save money, as if they are not concurrent then the cache reuse will work

Yes, if they are not concurrent then the cache reuse will work.

> we will need to provide semantics of serialization on that same key on the cache. And most importantly have the right UX to show a potential in progress execution for the same task with matching inputs

Yes, I think we point cached tasks to the original task already today. I think something like that would work for the UI.

> if the timeout is less than the total execution time, then it will only wait for the timeout and if cache is still not available, it will proceed

Yes, I think have a timeout would work here. Is the timeout for waiting on an existing execution with the same serialization to *finish*? Also, if there's no existing execution with the same serialization at the beginning of the execution, then it would just continue executing and not wait for the timeout to expire right? Finally, if the existing execution is not completed after the timeout, then the current execution would just execute?

**Comment by kumare3** [comment]
Awesome, just one clarification

> Yes, I think have a timeout would work here. Is the timeout for waiting on an existing execution with the same serialization to finish? Also, if there's no existing execution with the same serialization at the beginning of the execution, then it would just continue executing and not wait for the timeout to expire right? Finally, if the existing execution is not completed after the timeout, then the current execution would just execute?

Yes you got it right. The timeout is just to ensure liveness, without relying on our own failure detector. This is especially useful in disastrous scenarios where we lose a K8s cluster or data corruption.

**Comment by kumare3** [comment]
@michaels-lyft Also one more question, what would you say is the relative priority for this? Do you have an estimated timeline? Do you think anyone from your side can help or partner up?

**Comment by slai** [comment]
I generally like this idea, and conceptually, Flyte just needs to check the list of running tasks before starting a new task, and if an existing task with the same inputs (+ code version etc) is running, wait on that task as well. I can see people using this feature as a way to ensure singleton tasks too (or limited number).

I don't really understand why a timeout is needed. I also haven't looked into how Flyte works internally, but it seems like if multiple executions are waiting on that single task to complete, and that task fails or 'disappears' due to infra issues, then all of those executions should report the same status and fail in the same way. I would be wary of making this system too resilient to the point where it is unpredictable - caching is complex enough as it is :)

**Comment by kumare3** [comment]
@slai I agree, simplicity is Key. But, the timeout is needed, just because there are cases in which the system may be in a un-recoverable state - the cluster goes offline and hence there is no way to know that the task is no longer running etc. In this case - we will need to do active health-checking (usually implemented as Leases) to ensure that we can correctly unblock all the waiting tasks. 
A timeout, is like a a fail-safe. If you do not see an output within some max-time, then just run the execution. Granted this, means the timeout has to be correctly specified. But, it does not need to be exactly matching the execution time for the task, as in 99.99% cases things will work and everything will just continue, but for the 0.01% cases, the timeout will kick in and it can easily be 1.5x-2x the conservative estimate for the execution?

**Comment by michaels-lyft** [comment]
@kumare3 This is a nice to have feature that would allow us to kick off workflows concurrently without manually managing the kickoff time. But I can partner with you to work on the feature if you could point out where I can look into to make the change.

**Comment by kumare3** [comment]
@michaels-lyft sorry for the delay in responding. It would be great if you could partner on this. This should involve a short spec (one pager) and follow up with the implementation. I can help with the implementation in some places.

Steps
1. Add/update an API in Datacatalog that allows for conditionally adding a new task-execution in not ready status, thus fetching a dataset - should return this status as well and the time when the execution is set to expire
2. Use this API in Flytepropeller and pass the right event to FlyteAdmin

@michaels-lyft - Part 1 is a completely isolate piece of work in - https://github.com/flyteorg/datacatalog. Once this is done, I can work on 2 and add it pretty quickly.
Can you help with Part 1?
The gRPC spec for the service is available here - https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/datacatalog/datacatalog.proto

**Comment by kumare3** [comment]
@tnsetting what do you think about this feature?

**Comment by milton0825** [comment]
@kumare3 I can help with some of the implementation

**Comment by tnsetting** [comment]
@kumare3 sorry for the late response. I think this feature is good enhancement for datacatalog as we won't run the same task execution even for concurrent workflow executions. 

Just one thing regarding the timeout. In case of a bit long running jobs like 24 hours, does that implies other tasks have to wait until the timeout or we can refresh the timeout in some intervals like every hour?

**Comment by michaels-lyft** [comment]
@kumare3 Thanks for sharing the steps! In terms of timeline, I might be able to help out sometime in May or June. Does that work for you?

**Comment by kumare3** [comment]
@michaels-lyft actually @milton0825 from lyft may work on this in the next couple weeks. We can all collaborate to get it over the line

**Comment by milton0825** [comment]
Another option would be introduce pool (like airflow).
For example, we have a pool `pool_a` with size of 5, we can have at most 5 tasks in the pool running in parallel. To assign a task to a pool, we can update the flytekit interface with:
```
@task(pool='hahah')
```
and in this case for ETA, we have a pool size of 1 to ensure that we have one task running at a time. And of course the task need to have a timeout.

**Comment by michaels-lyft** [comment]
@milton0825 I'm not sure if this would work for our use case since we want to let multiple tasks run at the same time if they have different inputs.

**Comment by kanterov** [comment]
Can it be beneficial to have a plugin-specific solution? Without relying on any specifics, there is a problem with tasks dying that requires leases, and one approach to deal with that can be timeouts. On the other side, if we make this task-type-specific, there are ways to see if a task is still running, for instance, by polling pod for container tasks, or external systems, for instance for Presto plugin.

**Comment by milton0825** [comment]
@michaels-lyft I see. Yeah won't solve your case then if inputs are considered for determining concurrency.

**Comment by milton0825** [comment]
Created a one-pager draft: 
https://docs.google.com/document/d/1Oq8f0yVoHMHDCUbxhu9V-4dDLzbbM4YWcnNNscEnYkM/edit?usp=sharing

**Comment by kumare3** [comment]
@milton0825 left a few comments.
@kanterov can you please explain?

**Comment by kanterov** [comment]
@kumare3 we want to deduplicate task executions, which can be done by storing information if the task is running or not in the data catalog. Once flytepropeller instance starts to execute a task, it will update the status in the data catalog. During execution flytepropeller updates execution status in flyteadmin, which isn't 100% reliable. If flytepropeller instance dies, it isn't only going to fail to update status in flyteadmin, but also fail to notify data catalog. One way to see it is by implementing a leasing mechanism, where owning a lease will be an exclusive right to start execution. A simple implementation can grant a lease at the moment when "running" status is initially posted to the data catalog, and have a timeout. A drawback of that is that in the case flytepropeller instance failure, we are going to have to wait for a timeout. A way to decrease it is to have shorter timeouts and renew leases. 

The idea I brought is an alternative to managing leases, where instead of that, plugin implementation can get information about running tasks, and act accordingly. And then it's up to the plugin to decide how to handle that, for instance, either start speculative execution or wait for existing execution to finish. We don't want to rely on flyteadmin to know if a task is running. Polling output location can be problematic for the same reason. If we consider the "container" plugin as an example, if it can obtain the pod name, it can start monitoring the status of this pod instead. This way we don't have to specify timeouts or implement more complicated leasing.

**Comment by kumare3** [comment]
@kanterov I agree with the potential delay in processing in case of catastrophic failure. this happens only if we lose the etcD cluster (not if we lose propeller as it will recover from etcD). 
Lets break down the problem,
we want this to be done only in the case of catalog caching. Ideally if you are using flyte the right way, there is not need to concurrency control as data is guaranteed to be created in an independent/idempotent way.
Datacatalog is only available to the node - by design, to simplify how the memoization is done. Leaking this to the task increases complexity. But, I do see that speculation could be interesting.
I think what we are hinting at is leasing, but not really leasing, just a simple way to control the flow. Because in the worst case if 2 executions run, the consequences are not dramatic - we basically get 2 executions, which we are today. I dont think this affects correctness and we should not make this as a concurrency control - as that might be problematic. This is in reality a cost optimization

**Comment by kanterov** [comment]
@kumare3 Thanks for the explanation. I didn't know that propeller stores execution state in etcd. This makes the risk less relevant than I expected. It will be great if we can have good defaults so users don't have to tune timeouts, because being able to tune seems to be less relevant.

**Comment by michaels-lyft** [comment]
Hi @kumare3 Just wondering if there's an update on this issue?

**Comment by kumare3** [comment]
@milton0825 do you have an update?

**Comment by milton0825** [comment]
Unfortunately haven't been able to find time to work on this. Feel free to grab it.

**Comment by kumare3** [comment]
@hamersaw has started working on this
